Enabling empty object as a valid multiSelect hash

vdua opened the pull request:
## Description
Enabling empty object to be treated as a valid multiSelectHash. This will enable the following expression
`if(true(), {}, 'something else')`. Without this change the expression would be

```
if(true(), `{}`, 'something else')
```

There is one thing to note that the expressions like `a.{}` will also get supported in the grammar. Though they do not make any sense and we could throw an exception in the parser for them but I decided to keep them as they are not doing any harm.

## Related Issue
#18 

## Tasks
<!--- These tasks need to be done in order to get the PR merged, please mark with `x` if done or if they are not applicable to you or the change -->

- [x] I have signed the [CLA](http://adobe.github.io/cla.html).
- [x] I have written tests and verified that they fail without my change.

vdua commented on README.md:
```diff
@@ -27,7 +27,7 @@ If an expression fails to evaluate, follow these steps to debug:
 # populate src/test/debug.txt with the expression to test
 > cd antlr
-> antlr4 JSONFormula.g4
+> antlr JSONFormula.g4
```
It was an accidental change. Will revert

vdua commented:
@JohnBrinkman I am not sure do we need to generate the JavaScript Parser/Visitors for antlr4 grammar. I think we should parse the test cases with that grammar to ensure that the grammar stays correct. WDYT ?

JohnBrinkman commented:
>  I think we should parse the test cases with that grammar to ensure that the grammar stays correct. WDYT ?

Yes, parsing the test cases to validate the grammar makes sense.

vdua commented on src/test/jmespath/syntax.json:
```diff
@@ -414,11 +414,6 @@
         "expression": "a{foo: 0}",
         "error": "syntax"
       },
-      {
```
This is not an error case anymore. It has been moved https://github.com/adobe/json-formula/pull/22/commits/f928823ab186864eb08da26f9d3c72b7b255a0b5#diff-247931607bb8045c1f5499881916139eb8e02ccc27105b6209a0b397b6f6e907R438

vdua commented:
@JohnBrinkman added the support for parsing tests using antlr4 grammar. Currently there are some manual steps involved (added those in the Readme). Also the following expressions are not correct as per the antlr4 grammar

In Slice.json
```
foo[2:a:3]
```

In Syntax.json
```
slice expected colon or rbracket ([:@])
slice has too many colons ([:::])
slice expected number ([:@:])
Literal char not escaped (foo[?bar==`["foo`bar"]`])
Unknown comparator (foo[?bar^baz])
```

Our code understands this function but the parser generated from antlr4 grammar throws an exception.

JohnBrinkman commented:
@vdua 

> Our code understands these expressions but the parser generated from antlr4 grammar throws an exception.

I assume the antlr parser is correct?  i.e. we should make the JS code more strict?

JohnBrinkman reviewed (approved):
👍

JohnBrinkman commented:
@vdua Sadly ... I merged Arneh's changes first and now this PR has conflicts

vdua commented:
@JohnBrinkman I have resolved the conflicts. We added the support for expression in the brackets which the antlr parser generated from the grammar is not able to parse. We may need to update the grammar file for that use case. @arnehjain updated the grammar, but I think there are some rules and the auto generated parser is getting confused in picking

cc @arnehjain 

```
foo[2:a:3]
[:@]
[:::]
[:@:]
foo[?bar==`["foo`bar"]`
foo[?bar^baz]
```

JohnBrinkman commented:
Strictly speaking, we don't have to implement `^` as a new operator, as there is a `'power()` function.
I'm not opposed to adding it, but it is a change we need to pass by the working group.

vdua commented:
@JohnBrinkman That is already implemented in the code. If is not approved, then we should remove it.

vdua commented:
@JohnBrinkman @arnehjain I have removed the support for more than 2 cols in slice expressions. `[:::]` will not be supported but `[::]` will still work as per jmespath

Updated the grammar to allow expressions in slices.

There are two changes pending. 

* I have not removed `^` from our parser.
* The expression below is supported as per our antlr4 grammar file but not by our parser. 
It will take a little bit more time than I thought. Can we merge the changes now and decide whether we want to support this or not ?

```
`["def`xyz"]`
``` 

I recommend merging this PR and discuss and fix the above two issues separately #23 #24
